libc/src/setjmp/x86_64/longjmp.cpp

```diff
@@ -8,35 +8,83 @@
 
 #include "src/setjmp/longjmp.h"
 #include "include/llvm-libc-macros/offsetof-macro.h"
+#include "src/__support/OSUtil/io.h"
 #include "src/__support/common.h"
 #include "src/__support/macros/config.h"
+#include "src/setjmp/checksum.h"
+#include "src/stdlib/abort.h"
 
 #if !defined(LIBC_TARGET_ARCH_IS_X86_64)
 #error "Invalid file include"
 #endif
 
 namespace LIBC_NAMESPACE_DECL {
 
+#if LIBC_COPT_SETJMP_ENABLE_FORTIFICATION
+extern "C" [[gnu::cold, noreturn]] void __libc_jmpbuf_corruption() {
+  write_to_stderr("invalid checksum detected in longjmp\n");
+  abort();
+}
+#define LOAD_CHKSUM_STATE_REGISTERS()                                          \
+  asm("mov %0, %%rcx\n\t" ::"m"(jmpbuf::value_mask) : "rcx");                  \
+  asm("mov %0, %%rdx\n\t" ::"m"(jmpbuf::checksum_cookie) : "rdx");
+
+#define RESTORE_REG(DST)                                                       \
+  "movq %c[" #DST "](%%rdi), %%rax\n\t"                                        \
+  "movq %%rax, %%" #DST "\n\t"                                                 \
+  "xor %%rcx, %%" #DST "\n\t"                                                  \
+  "mul %%rdx\n\t"                                                              \
+  "xor %%rax, %%rdx\n\t"                                                       \
+  "rol $%c[rotation], %%rdx\n\t"
```

Comment:
If I've read this correctly, the checksum algorithm updates the 64-bit checksum with each extra 64-bit word by means of

1. multiplying the checksum by the _masked_ version of the word, generating a 128-bit product
2. reducing back to 64 bits by XORing the two halves of the product
3. rotating the product left by 13 bits

Is this a well-known lightweight checksum algorithm used in other places, or is it made up entirely for this patch? Is there anything written down anywhere on why it's a good choice for this somewhat security-related application?

If this feature is later implemented on other architectures (say AArch64), would you recommend that implementors follow the identical algorithm, or that they choose something arbitrary that has good performance on that architecture?

Reply:
It is a variant implementation of the simple multiply-shifting hash. I would say it is not crypto-safe but we only need to check the integrity rather than hiding information. It is relatively generic and fast, which is the main reason that I choose it. For aarch64, maybe `pmull` or `aesni` can also be a choice?

Reply:
See also https://github.com/tkaitchuck/aHash/blob/7d5c661a74b12d5bc5448b0b83fdb429190db1a3/src/fallback_hash.rs#L65C1-L93C84

Reply:
Checking integrity is also a well studied cryptographic problem, just as much as encryption to hide information. If the aim is to protect against an attacker, the question is: how easy is it to make a controlled change to the `jmp_buf` and update the hash value so that the check passes?

The switching between XOR and multiplication makes it not immediately obvious, but I'm sure a cryptanalyst would be able to find a break.

I do understand that you're trying to balance having _some_ protection against not losing too much speed. So I'm not saying you should switch over to AES and HMAC. But I do think you should write down your reasoning in a comment alongside the algorithm itself.